src/puppetlabs/puppetdb/cli/benchmark.clj

```diff
@@ -472,27 +429,148 @@
           (<! message-timeout)
           (recur))))))
 
-(defn benchmark-main
-  [& args]
+(defn random-hosts
+  [n pdb-host catalogs reports facts]
+  (let [random-entity (fn [host entities]
+                        (some-> entities
+                                rand-nth
+                                (assoc "certname" host)))]
+    (for [i (range n)]
+      (let [host (str "host-" i)]
+        {:host host
+         :catalog (when-let [cat (random-entity host catalogs)]
+                    (update cat "resources"
+                            (partial map #(update % "tags"
+                                                  conj
+                                                  pdb-host))))
+         :report (random-entity host reports)
+         :factset (random-entity host facts)}))))
+
+(defn populate-queue
+  "Fills queue with host entries.  Stops if mq is closed."
+  [mq numhosts run-interval pdb-host catalogs reports facts]
+  (println-err (trs "Populating queue in the background"))
+  (doseq [host (random-hosts numhosts pdb-host catalogs reports facts)
+          :while (>!! mq (assoc host :lastrun (rand-lastrun run-interval)))]
+    true)
+  (println-err (trs "Finished populating queue")))
+
+(defn cycle-commands [numhosts run-interval rand-perc simulation-threads
+                      rate-monitor-ch command-send-ch
+                      pdb-host catalogs reports facts]
+  "Populates the message queue with messages, and then cycles through
+  those messages, sending and altering each one every time it
+  reappears at the head of the queue.  Returns a stop function of no
+  arguments that can be used to effect an orderly shutdown."
+  ;; The stop function is also registered as a shutdown hook (thread),
+  ;; and is set up as a oneshot, protected by an atomic boolean.  All
+  ;; of the work that needs to be completed on a SIGINT, etc. is
+  ;; coordinated from that hook via the stop function because after a
+  ;; signal, only daemon threads, shutdown hooks, and finalizers will
+  ;; still run, and daemon threads will only run until all of the
+  ;; hooks and finalizers are finished.
+  (let [preftmp (get-path (or (System/getenv "TMPDIR")
```

Comment:
I'd mentioned it in load_testing_tool.markdown; perhaps I should also add it to the --help.